### PR TITLE
 Set OAuth2Client idle when Nessie client is idle

### PR DESCRIPTION
**Pull request by adutra**
This PR introduces 2 mechanisms to mitigate the situation where the Nessie client is idle but the OAuth2 client keeps consuming resources:

1. An idle interval after which the tokens are not refreshed anymore. When the token is needed again, if it's expired, it will be renewed from step 1.
2. A keep alive interval after which the background thread is stopped. If a refresh is needed again, a new thread will be spawned.

It was hard to create tests for these new features. I will see if I can improve that more.

### CODE REVIEW

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java [hunk 1]
```diff
@@ -162,6 +164,33 @@ void testOAuth2ClientInitialRefreshToken() {
     }
   }
 
+  @Test
+  void testOAuth2ClientIdle() throws InterruptedException {
```
Comment (adutra) [inline api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:168]: The test is admittedly a little lame, and takes 15 seconds. I wonder if it's necessary to keep it around. We could disable it maybe and only run when necessary.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 2]
```diff
@@ -65,15 +65,18 @@ class OAuth2Client implements OAuth2Authenticator, Closeable {
   private final Duration defaultAccessTokenLifespan;
   private final Duration defaultRefreshTokenLifespan;
   private final Duration refreshSafetyWindow;
+  private final Duration idleInterval;
   private final boolean tokenExchangeEnabled;
   private final HttpClient httpClient;
-  private final ScheduledExecutorService executor;
+  /* Visible for testing. */ final ScheduledExecutorService executor;
```
[the remark this answers is not in the document]
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:71]: That would bring Guava into compile scope, are we OK with that?

[the remark this answers is not in the document]
Comment (dimas-b) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:71]: I initially thought a compile-only dependency would be fine, but now I'm thinking that it is likely to cause compilation warnings in downstream projects... I'm fine keeping the comment.

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:168]: The idea of injecting a Clock is nice, I will look into that. However, it will imho help mostly with unit tests. Here, just the clock won't be enough because we also need to control when the executor is going to run the tasks, and even worse, we can't really control the token expiration times that Keycloak creates.

Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:168]: I will try however to "shrink" the test execution time by making the overall delays more aggressive.

Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: The unit test `TestOAuth2Client#testBackgroundRefresh`, which mocks both the executor + keycloak, now checks the sleeping flag.

The only extra thing we test here is the executor shrinking its core thread pool to zero.

I managed to reduce this test to ~5 secs btw.

Otherwise yes we could mock Keycloak but that would lead to something like Quarkus `OidcWiremockResource` which seems a lot of work for little gain, wdyt?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 3]
```diff
@@ -65,15 +66,19 @@ class OAuth2Client implements OAuth2Authenticator, Closeable {
...
   private final boolean shouldCloseExecutor;
   private final ObjectMapper objectMapper;
   private final CompletableFuture<Void> started = new CompletableFuture<>();
+  /* Visible for testing. */ final AtomicBoolean sleeping = new AtomicBoolean();
+  private final Supplier<Instant> clock;
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:77]: Why's this a supplier instead of a `Clock`? (We do have that threeten-extra package which provides a mutable clock impl).

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java [hunk 4]
```diff
@@ -162,6 +164,33 @@ void testOAuth2ClientInitialRefreshToken() {
...
+      client.start();
+      assertThat(client.sleeping).isFalse();
+      // after refresh tokens, will enter sleep mode (idle interval was 1 sec)
+      await().until(client.sleeping::get);
+      // thread will exit after 1 sec (keep alive interval 1 sec)
+      await().until(() -> ((OAuth2TokenRefreshExecutor) client.executor).getPoolSize() == 0);
+      // access token is now expired; calling authenticate() will wake up client, fetch new tokens
+      // immediately (on main thread), then schedule refresh, which will in turn spawn a new thread
+      AccessToken accessToken = client.authenticate();
+      assertThat(client.sleeping).isFalse();
```
Comment (dimas-b) [inline api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:188]: This is potentially flaky as there's no barrier that prevents the client from sleeping again and the test execution could be delayed.

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (dimas-b) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: Could we mock at the OAuth2 HTTP/REST level?... e.g. delegating `fetchNewTokens()` and similar methods to an impl. controlled by the test?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 5]
```diff
@@ -165,35 +169,57 @@ public void close() {
...
 
-  private CompletionStage<Tokens> renewTokens() {
-    return currentTokensStage
-        // try refreshing the current tokens
-        .thenApply(this::refreshTokens)
-        // if that fails, try fetching brand-new tokens
-        .exceptionally(error -> fetchNewTokens())
-        .whenComplete(this::log);
+  /** Visible for testing. */
+  boolean idle(Instant now) {
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:214]: Can you rename this to `shouldSleep`?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 6]
```diff
@@ -165,35 +169,57 @@ public void close() {
...
+    return Duration.between(lastAccess, now).compareTo(idleInterval) > 0;
+  }
+
+  private void wakeUp(Instant now) {
+    scheduleOrExecuteTokensRenewal(getCurrentTokens(), now, Duration.ZERO);
   }
 
-  private void log(Tokens tokens, Throwable error) {
+  private void log(Tokens redacted, Throwable error) {
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:222]: Should the `redacted` parameter go away? it's unused.

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: That's what `TestOAuth2Client` does.

File: api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java [hunk 7]
```diff
@@ -156,6 +156,30 @@ public final class NessieConfigConstants {
...
+   * href="https://en.wikipedia.org/wiki/ISO_8601#Durations">ISO-8601 duration</a>.
+   */
+  public static final String CONF_NESSIE_OAUTH2_IDLE_INTERVAL =
+      "nessie.authentication.oauth2.idle-interval";
+
+  public static final String DEFAULT_KEEP_ALIVE_INTERVAL = "PT30S";
+
+  /**
+   * Config property name ({@value #CONF_NESSIE_OAUTH2_KEEP_ALIVE_INTERVAL}) for the OAuth2
+   * authentication provider. The keep alive interval to use; if the OAuth2 provider background
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java:175]: I think the description should be different - as it's from a user's perspective.

Form a user's perspective it's the time the background thread stays around to speed up token-refreshes.
It's rather the time the Nessie client will proactively refresh tokens.

Actually, the name of the parameter should also be something around that meaning and not "keep alive" - maybe "idle client token refresh timeout" or so (also not a brilliant idea...).

Should mention that setting this to a "too high" value _could_ is not a good way (thread leak, memory leak - if clients are re-created).

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 4 above)
Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:188]: Yes, at least in theory. I am leaning in favor of deleting this test. We don't need to test the keep alive stuff, it's not our code.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2TokenRefreshExecutor.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+import java.time.Duration;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class OAuth2TokenRefreshExecutor extends ScheduledThreadPoolExecutor {
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2TokenRefreshExecutor.java:26]: this class can be `final` (it's inner classes as well)

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (dimas-b) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: So can we do the sleep test there?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 9]
```diff
@@ -165,35 +169,57 @@ public void close() {
...
+      renewTokens();
+    } else {
+      LOGGER.debug("Scheduling token refresh in {}", delay);
+      tokenRefreshFuture =
+          executor.schedule(this::renewTokens, delay.toMillis(), TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private void renewTokens() {
+    CompletionStage<Tokens> oldTokensStage = currentTokensStage;
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:202]: Nit: Should `oldTokansStage` be inlined?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 10]
```diff
@@ -165,35 +169,57 @@ public void close() {
     }
   }
 
-  private void scheduleTokensRenewal(Tokens currentTokens) {
-    Instant now = Instant.now();
+  private void maybeScheduleTokensRenewal(Tokens currentTokens) {
+    Instant now = clock.get();
+    if (idle(now)) {
+      sleeping.set(true);
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:175]: Probably unlikely, but I suspect this update can race with the conditional update in `authenticate()`. A user (indirectly) calls `authenticate()` - that thread gets stuck after the CAS - this function runs and sets sleeping back to true - and then the user thread resumes. It feels, that it can wake-up twice.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 11]
```diff
@@ -165,35 +169,57 @@ public void close() {
...
-        // if that fails, try fetching brand-new tokens
-        .exceptionally(error -> fetchNewTokens())
-        .whenComplete(this::log);
+  /** Visible for testing. */
+  boolean idle(Instant now) {
+    return Duration.between(lastAccess, now).compareTo(idleInterval) > 0;
+  }
+
+  private void wakeUp(Instant now) {
+    scheduleOrExecuteTokensRenewal(getCurrentTokens(), now, Duration.ZERO);
```
Comment (snazy) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:219]: Can you inline this to the (only) call site?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 3 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:77]: Because it's more flexible than `Clock` and a `Clock` can be turned into a `Supplier<Instant>` with `clock::instant`.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 6 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:222]: It makes it possible to use a method reference in `whenComplete(this::log)`, but why not.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 9 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:202]: IJ gives a warning: "Non-atomic operation on volatile field 'currentTokensStage'".

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 11 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:219]: I thought having a method would better self-document the intent, but sure. We can also remove `shouldSleep` in fact.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 6 above)
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:222]: Ah, NVM then

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 9 above)
Comment (snazy) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:202]: Oh, interesting.

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 10 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:175]: > this function runs

In which thread? I don't think that's possible: if `sleeping` was set to true, this means that the background thread pool is idle, and there is no refresh task scheduled for later execution.

It could happen though, that a user thread enters `authenticate()`, resets `sleeping` to false, then renew the tokens (which involves network IO), then gets stuck for several seconds somewhere, then enters `maybeScheduleTokensRenewal`, then resets `sleeping` to true again. But that's not a pathological code path.

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: But... we already do: 

https://github.com/projectnessie/nessie/blob/a77c21ed9ef70d607834401885d738b59939af4d/api/client/src/test/java/org/projectnessie/client/auth/oauth2/TestOAuth2Client.java#L122-L130

Is there anything else you would like to test in that test class?

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 4 above)
Comment (adutra) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:188]: Alright, no objections so far: I'm removing this integration test.

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java [hunk 12]
```diff
@@ -73,7 +73,7 @@ static void setUpKeycloak() {
         URI.create(KEYCLOAK.getAuthServerUrl() + "/realms/master/protocol/openid-connect/token");
     Keycloak keycloakAdmin = KEYCLOAK.getKeycloakAdminClient();
     master = keycloakAdmin.realms().realm("master");
-    updateMasterRealm(10, 15);
+    updateMasterRealm(5, 10);
```
Comment (adutra) [inline api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java:76]: I thought we could keep these changes though, it's nice to have a shorter test execution time.

File: api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java (hunk 7 above)
Comment (dimas-b) [reply api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java]: `CONF_NESSIE_OAUTH2_IDLE_INTERVAL` -> `CONF_NESSIE_OAUTH2_PREEMPTIVE_TOKEN_REFRESH_IDLE_TIMEOUT` ? The "idle" part is a hint that the preemptive refresh will not happen when the client is idle.

`CONF_NESSIE_OAUTH2_BACKGROUND_THREAD_IDLE_TIMEOUT` - LGTM.

Comment (dimas-b) [reply api/client/src/main/java/org/projectnessie/client/NessieConfigConstants.java]: Also, I think we can keep the constant names shorter as they are not user-visible, but have descriptive property names (values of those constants).

File: api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java (hunk 1 above)
Comment (dimas-b) [reply api/client/src/intTest/java/org/projectnessie/client/auth/oauth2/ITOAuth2Client.java]: Hmmm.... I must have missed it or got confused :facepalm:

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java [hunk 13]
```diff
@@ -83,28 +88,29 @@ class OAuth2Client implements OAuth2Authenticator, Closeable {
...
+    currentTokensStage.thenAccept(this::maybeScheduleTokensRenewal);
   }
 
   @Override
   public AccessToken authenticate() {
+    Instant now = clock.get();
+    lastAccess = now;
+    if (sleeping.compareAndSet(true, false)) {
+      LOGGER.debug("Waking up...");
+      scheduleOrExecuteTokensRenewal(getCurrentTokens(), now, Duration.ZERO);
```
Comment (dimas-b) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:112]: if this throws, should we set `sleeping` back to `true` to force a renewal on the next `authenticate()` call?

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2TokenRefreshExecutor.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      Thread thread = new TokenRefreshThread(r);
+      thread.setDaemon(true);
+      return thread;
+    }
+  }
+
+  private static final class TokenRefreshThread extends Thread {
+
+    TokenRefreshThread(Runnable r) {
+      super(r, "oauth2-token-refresh");
```
Comment (dimas-b) [inline api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2TokenRefreshExecutor.java:49]: nit: `nessie-client-oauth2-token-refresh` maybe?

File: api/client/src/test/java/org/projectnessie/client/auth/oauth2/TestOAuth2Client.java [hunk 15]
```diff
@@ -115,6 +118,16 @@ void testBackgroundRefresh() throws Exception {
...
         assertThat(client.getCurrentTokens()).isInstanceOf(RefreshTokensResponse.class);
+
+        // emulate executor running the scheduled renewal task and detecting that the client is idle
+        // after 30+ seconds of inactivity
+        i.set(i.get().plusSeconds(31));
+        currentRenewalTask.get().run();
+        assertThat(client.sleeping).isTrue();
+
+        // should exit sleeping mode on next authenticate() call
+        client.authenticate();
```
Comment (dimas-b) [inline api/client/src/test/java/org/projectnessie/client/auth/oauth2/TestOAuth2Client.java:129]: Do you think we could add a test case for task scheduling failures? Those may happen with user-supplied executors (e.g. short queue), I guess... WDYT?

**Review by dimas-b — COMMENTED** [review]
LGTM overall. A few polishing comments :)

File: api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java (hunk 13 above)
Comment (adutra) [reply api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java:112]: The only way I can image where this would throw is if `executor.schedule` throws `RejectedExecutionException`. I prefer for this exception to propagate in fact, as the OAuth2 client is probably left in an usable state.

But you are remark is excellent; and I agree that we might want to revisit this and strengthen things a bit: e.g. if `executor.schedule` throws `RejectedExecutionException` inside the bakground thread, the error will be swallowed here:

https://github.com/projectnessie/nessie/blob/98448b542fd8c9f2f69a6dbeb265c3d94c6bf21c/api/client/src/main/java/org/projectnessie/client/auth/oauth2/OAuth2Client.java#L212

That's indeed bad and the client would be left in an unrecoverable state for sure. I think we should fail the `currentTokensStage` in that case, so that the exception gets propagated to client code when the user calls `authenticate()`.

Do you mind if I do this in another PR? As it's only partially related to sleeping/waking up.

File: api/client/src/test/java/org/projectnessie/client/auth/oauth2/TestOAuth2Client.java (hunk 15 above)
Comment (adutra) [reply api/client/src/test/java/org/projectnessie/client/auth/oauth2/TestOAuth2Client.java:129]: Yes, good idea. Do you mind if I tackle the scheduling failures stuff in a follow-up PR?

**Review by dimas-b — APPROVED** [review]
